src/drivers/intel/cavs/ssp.c

```diff
@@ -345,102 +346,57 @@ static inline int ssp_set_config(struct dai *dai,
-	}
+	/* MCLK config */
+	/* searching the smallest possible mclk source */
+	for (i = ARRAY_SIZE(ssp_freq) - 1; i >= 0; i--) {
+		if (config->ssp.mclk_rate > ssp_freq[i].freq)
+			break;
 
-	if (F_38400_kHz % config->ssp.bclk_rate == 0) {
-		mdiv = F_38400_kHz / config->ssp.bclk_rate;
-	} else {
-		trace_ssp_error("ssp_set_config() error: 38.4MHz / %d Hz BCLK not divisable",
-				config->ssp.bclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-#elif defined(CONFIG_CANNONLAKE)
-	if (!config->ssp.mclk_rate || config->ssp.mclk_rate > F_24000_kHz) {
-		trace_ssp_error("ssp_set_config() error: "
-				"invalid MCLK = %d Hz (valid < 24000kHz)",
-				config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-	if (!config->ssp.bclk_rate ||
-	    config->ssp.bclk_rate > config->ssp.mclk_rate) {
-		trace_ssp_error("ssp_set_config() error: "
-				"BCLK %d Hz = 0 or > MCLK %d Hz",
-				config->ssp.bclk_rate, config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
+		if (ssp_freq[i].freq % config->ssp.mclk_rate == 0) {
+			clk_index = i;
+			clk_set = 1;
+		}
 	}
 
-	if (F_24000_kHz % config->ssp.mclk_rate == 0) {
-		mdivr_val = F_24000_kHz / config->ssp.mclk_rate;
+	if (clk_set) {
+		mdivc |= MCDSS(ssp_freq[clk_index].enc);
+		mdivr_val = ssp_freq[clk_index].freq / config->ssp.mclk_rate;
 	} else {
-		trace_ssp_error("ssp_set_config() error: 24.0MHz / %d Hz MCLK not divisable",
+		trace_ssp_error("ssp_set_config() error: MCLK %d",
 				config->ssp.mclk_rate);
 		ret = -EINVAL;
 		goto out;
 	}
 
-	if (F_24000_kHz % config->ssp.bclk_rate == 0) {
-		mdiv = F_24000_kHz / config->ssp.bclk_rate;
-	} else {
-		trace_ssp_error("ssp_set_config() error: 24.0MHz / %d Hz BCLK not divisable",
-				config->ssp.bclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-#else
-	if (!config->ssp.mclk_rate || config->ssp.mclk_rate > F_24576_kHz) {
-		trace_ssp_error("ssp_set_config() error: "
-				"invalid MCLK = %d Hz (valid < 24567kHz)",
-				config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-	if (!config->ssp.bclk_rate ||
-	    config->ssp.bclk_rate > config->ssp.mclk_rate) {
-		trace_ssp_error("ssp_set_config() error: "
-				"BCLK %d Hz = 0 or > MCLK %d Hz",
-				config->ssp.bclk_rate, config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-	if (F_24576_kHz % config->ssp.mclk_rate == 0) {
-		/* select Audio Cardinal clock for MCLK */
-		mdivc |= MCDSS(1);
-		mdivr_val = F_24576_kHz / config->ssp.mclk_rate;
-	} else if (config->ssp.mclk_rate <= F_19200_kHz &&
-		   F_19200_kHz % config->ssp.mclk_rate == 0) {
-		mdivr_val = F_19200_kHz / config->ssp.mclk_rate;
-	} else {
-		trace_ssp_error("ssp_set_config() error: MCLK %d",
-				config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
+	/* BCLK config */
+	/* searching the smallest possible bclk source */
+	clk_set = 0;
+	for (i = ARRAY_SIZE(ssp_freq) - 1; i >= 0; i--) {
+		if (config->ssp.bclk_rate > ssp_freq[i].freq)
+			break;
+
+		if (ssp_freq[i].freq % config->ssp.bclk_rate == 0) {
+			clk_index = i;
+			clk_set = 1;
+		}
```

Comment:
this doesn't select the smallest possible source, e.g. for IceLake the frequencies listed are 19.2 and 38.4, so you'd end-up taking 38.4.

Reply:
Now for ICELAKE there are 3 frequences: 96, 38.4 , 24.576 MHz. I think that algorithm choose the smallest possible freq. I'm searching using for loop:
```
for (i = MAX_SSP_FREQ_INDEX; i >= 0; i--) {
		if (config->ssp.mclk_rate > ssp_freq[i].freq)
			break;

		if (ssp_freq[i].freq % config->ssp.mclk_rate == 0) {
			clk_index = i;
			clk_set = 1;
		}
	}
```
It's going through the highest to the lowest frequency. And checking ```if (ssp_freq[i].freq % config->ssp.mclk_rate == 0)``` condition.I use break statement only if ```config->ssp.mclk_rate > ssp_freq[i].freq```